lib/Bindings/Python/circt/dialects/_rtl_ops_ext.py

```diff
@@ -1,16 +1,121 @@
+               parameters={},
+               loc=None,
+               ip=None):
+    self.module = module
+    self.instance_name = StringAttr.get(name)
+    self.module_name = FlatSymbolRefAttr.get(StringAttr(self.module.name).value)
+    self.parameters = DictAttr.get(parameters)
+    self.loc = loc
+    self.ip = ip
+    self.instance = None
+    self.arg_values = {}
+    self.result_values = {}
+
+    arg_names = ArrayAttr(self.module.attributes["argNames"])
+    result_names = ArrayAttr(self.module.attributes["resultNames"])
+
+    def string_value(attr):
+      return StringAttr(attr).value
+
+    self.arg_names = list(map(string_value, arg_names))
+    self.arg_names_set = frozenset(self.arg_names)
+    self.result_names = list(map(string_value, result_names))
+    self.result_names_set = frozenset(self.result_names)
+
+    for (name, value) in input_port_mapping.items():
+      if name in self.arg_names_set:
+        self.arg_values[name] = value
+      else:
+        raise AttributeError(f"unknown input port name {name}")
+
+    self.maybe_build()
+
+  def maybe_build(self):
+    # Ensure all the arg values have been stored.
+    if len(self.arg_values) < len(self.arg_names):
+      return
+
+    # Get the stored arg values in the correct order.
+    arg_values = [self.arg_values[name] for name in self.arg_names]
+
+    # Lazily import InstanceOp to avoid cyclic dependencies.
+    from ._rtl_ops_gen import InstanceOp
+
+    # Actually build the InstanceOp.
+    self.instance = InstanceOp(
+        self.module.type.results,
+        self.instance_name,
+        self.module_name,
+        arg_values,
+        self.parameters,
+        loc=self.loc,
+        ip=self.ip,
+    )
+
+  def __getattr__(self, name):
+    # Check for the attribute in the result name set.
+    if name in object.__getattribute__(self, "result_names_set"):
+      # Ensure the instance is fully instantiated.
+      instance = object.__getattribute__(self, "instance")
+      if not instance:
+        raise AttributeError("instance is not yet fully-defined")
+
+      # Return the instance result with this name.
+      result_names = object.__getattribute__(self, "result_names")
+      return instance.results[result_names.index(name)]
+
+    # If we fell through to here, the name isn't a result.
+    raise AttributeError(f"unknown output port name {name}")
+
+  def __setattr__(self, name, value):
+    # If we are actually setting an InstanceBuilder attribute, just do that.
+    if name in self.__slots__:
+      object.__setattr__(self, name, value)
+      return
+
+    # Check for the attribute in the arg name set.
+    if name in object.__getattribute__(self, "arg_names_set"):
+      # Store the value for building the InstanceOp.
+      arg_values = object.__getattribute__(self, "arg_values")
+      arg_values[name] = value
+
+      # Attempt to build the InstanceOp.
+      maybe_build = object.__getattribute__(self, "maybe_build")
+      maybe_build()
```

Comment:
Note that this is (roughly) implementing the second choice from point 1. in [this comment](https://github.com/llvm/circt/issues/905#issuecomment-817023205): it doesn't actually build an instance until all of its _inputs_ are known. We don't need to wait to know the full connectivity of every port, just the inputs.

Reply:
I haven't had the chance to read this thoroughly, but does this work in the presence of cycles? i.e. two modules which feed each other.

Reply:
Good question. My intuition is no... I will try it out though.

Reply:
Yep confirmed that this doesn't work in the current revision. It falls through to this scenario here: https://github.com/llvm/circt/pull/940/files#diff-45ef9b344a85b7561366d5f9457f7de8579b64d97164cf4e4adc21462b11ef1bR82.

That error wasn't actually tested in this patch, so I a test for this behavior in the most recent commit.

I'm thinking about how to handle cycles using the current, lazy approach, and I think I have an intuition for it. Should be able to add it to this PR or submit it as a follow up. If that's not working out, we can always go for the backedge-style approach, where the instance is initially buillt with dummy values that get swapped out as we go.